### PR TITLE
Include z slope cut (+ timing debugs) and change sum chi2

### PR DESCRIPTION
**Pull request by ebrondol**
BEGINRELEASENOTES
- Include cut in slope in z for CA building part to reduce number of combinatorics
- Change sum of chi2 from squared root of the squared sum to linear sum
- No big changes expected in eff/fakes
- Reduce CPU time
ENDRELEASENOTES

### CODE REVIEW

File: include/ConformalTracking.h [hunk 1]
```diff
@@ -77,6 +78,7 @@ class ConformalTracking : public Processor {
   // Track finding
   void buildNewTracks(UniqueKDTracks&, SharedKDClusters&, UKDTree&, Parameters const&, bool radialSearch = false,
                       bool vertexToTracker = true);
+  bool neighbourIsCompatible(const SKDCluster& neighbourHit, SKDCluster& seedHit);
```
Comment (andresailer) [inline include/ConformalTracking.h]: Both arguments can be `const`, right?

File: src/ConformalTracking.cc [hunk 2]
```diff
@@ -125,6 +124,7 @@ void ConformalTracking::registerParameters() {
   registerOutputCollection(LCIO::TRACKERHITPLANE, "DebugHits", "DebugHits", m_outputDebugHits, std::string("DebugHits"));
 
   registerProcessorParameter("DebugPlots", "Plots for debugging the tracking", m_debugPlots, bool(false));
+  registerProcessorParameter("DebugTiming", "Print out time recording", m_debugTime, bool(false));
```
Comment (andresailer) [inline src/ConformalTracking.cc]: what is "time recording"?

File: src/ConformalTracking.cc [hunk 3]
```diff
@@ -1257,9 +1354,30 @@ void ConformalTracking::buildNewTracks(UniqueKDTracks& conformalTracks, SharedKD
...
 }
 
+// Check that it the neighboour has the z slope is in the range
+bool ConformalTracking::neighbourIsCompatible(const SKDCluster& neighbourHit, SKDCluster& seedHit) {
+  double neighbourHit_r = sqrt(neighbourHit->getX() * neighbourHit->getX() + neighbourHit->getY() * neighbourHit->getY());
+  double seedHit_r      = sqrt(seedHit->getX() * seedHit->getX() + seedHit->getY() * seedHit->getY());
+  double distance       = neighbourHit_r - seedHit_r;
+  double deltaZ         = neighbourHit->getZ() - seedHit->getZ();
+  if (fabs(deltaZ / distance) > m_slopeZRange) {
+    streamlog_out(DEBUG7) << "- z condition conditions not met" << std::endl;
```
Comment (andresailer) [inline src/ConformalTracking.cc]:
```suggestion
    streamlog_out(DEBUG7) << "- z condition not met" << std::endl;
```

File: src/ConformalTracking.cc [hunk 4]
```diff
@@ -2876,15 +2995,15 @@ void ConformalTracking::runStep(SharedKDClusters& kdClusters, UKDTree& nearestNe
     } while (caughtException);
 
     stopwatch.Stop();
-    streamlog_out(DEBUG7) << "Step " << parameters._step << "buildNewTracks took " << stopwatch.RealTime() << " seconds"
-                          << std::endl;
+    if (m_debugTime)
+      std::cout << "Step " << parameters._step << " buildNewTracks took " << stopwatch.RealTime() << " seconds" << std::endl;
```
Comment (andresailer) [inline src/ConformalTracking.cc]: don't use `cout`

File: src/ConformalTracking.cc [hunk 5]
```diff
@@ -2876,15 +2995,15 @@ void ConformalTracking::runStep(SharedKDClusters& kdClusters, UKDTree& nearestNe
...
     stopwatch.Reset();
   }
   if (parameters._extend) {
     extendTracks(conformalTracks, kdClusters, nearestNeighbours, parameters);
     stopwatch.Stop();
-    streamlog_out(DEBUG7) << "Step " << parameters._step << "extendTracks took " << stopwatch.RealTime() << " seconds"
-                          << std::endl;
+    if (m_debugTime)
+      std::cout << "Step " << parameters._step << " extendTracks took " << stopwatch.RealTime() << " seconds" << std::endl;
```
Comment (andresailer) [inline src/ConformalTracking.cc]: don't use `cout`

File: src/ConformalTracking.cc [hunk 6]
```diff
@@ -1257,9 +1354,30 @@ void ConformalTracking::buildNewTracks(UniqueKDTracks& conformalTracks, SharedKD
...
+                            << " milli-seconds" << std::endl;
+    if (m_debugTime)
+      streamlog_out(DEBUG7) << " Time report: Total time for seed hit " << nKDHit << " = "
+                            << stopwatch_hit_total.RealTime() * 1000 << std::scientific << " milli-seconds" << std::endl;
   }
 }
 
+// Check that it the neighboour has the z slope is in the range
+bool ConformalTracking::neighbourIsCompatible(const SKDCluster& neighbourHit, SKDCluster& seedHit) {
+  double neighbourHit_r = sqrt(neighbourHit->getX() * neighbourHit->getX() + neighbourHit->getY() * neighbourHit->getY());
```
Comment (andresailer) [inline src/ConformalTracking.cc]: `kdClusters` already store the radius, not need to calculate it again.

File: src/ConformalTracking.cc [hunk 7]
```diff
@@ -931,10 +960,44 @@ void ConformalTracking::buildNewTracks(UniqueKDTracks& conformalTracks, SharedKD
     // radial distance. Check that they are at lower radius and further from the IP
     SharedKDClusters results;
     double           theta = kdhit->getTheta();
+    // Filter already if the neighboour is used, is on the same detector layer,
```
Comment (andresailer) [inline src/ConformalTracking.cc]:
```suggestion
    // Filter already if the neighbour is used, is on the same detector layer,
```

**Comment by ebrondol** [comment]
Efficiency for ttbar w/o and w/ overlay
ttbar_eff_vs_pt_comparison_commit_7c5c143
ttbar_ove_eff_vs_pt_comparison_commit_7c5c143
Fake rate for ttbar + overlay
ttbar_ove_fake_vs_pt_comparison_commit_7c5c143
ttbar_ove_fake_vs_theta_comparison_commit_7c5c143

File: src/ConformalTracking.cc [hunk 8]
```diff
@@ -1257,9 +1356,31 @@ void ConformalTracking::buildNewTracks(UniqueKDTracks& conformalTracks, SharedKD
...
+    if (m_debugTime)
+      streamlog_out(DEBUG7) << "  Time report: Sort best tracks took " << stopwatch_hit.RealTime() * 1000 << std::scientific
+                            << " milli-seconds" << std::endl;
+    if (m_debugTime)
+      streamlog_out(DEBUG7) << " Time report: Total time for seed hit " << nKDHit << " = "
+                            << stopwatch_hit_total.RealTime() * 1000 << std::scientific << " milli-seconds" << std::endl;
   }
 }
 
+// Check that it the neighbour has the z slope is in the range
```
Comment (andresailer) [inline src/ConformalTracking.cc]:
```suggestion
/// Check that the neighbour has the z slope in the required range compared to the seed
```

File: src/ConformalTracking.cc [hunk 9]
```diff
@@ -1257,9 +1356,31 @@ void ConformalTracking::buildNewTracks(UniqueKDTracks& conformalTracks, SharedKD
...
+bool ConformalTracking::neighbourIsCompatible(const SKDCluster& neighbourHit, const SKDCluster& seedHit,
+                                              const double slopeZRange) {
+  double distanceX = neighbourHit->getX() - seedHit->getX();
```
Comment (andresailer) [inline src/ConformalTracking.cc]:
```suggestion
  const double distanceX(neighbourHit->getX() - seedHit->getX());
```
etc.

**Pull request merged**